File: api/src/graphql/question.rs
use juniper::{graphql_object, FieldResult, GraphQLInputObject, GraphQLObject, ID};
use mongodb::bson::oid::{self, ObjectId};

use crate::context::Context;
use crate::models::{self, BaseModel};

#[derive(GraphQLObject)]
pub struct Question {
    answer: String,
    id: ID,
    question: String,
}

impl Question {
    pub fn from_model(model: &models::Question) -> Self {
        Self {
            answer: model.answer.clone(),
            id: model.id.to_string().into(),
            question: model.question.clone(),
        }
    }
}

#[derive(GraphQLInputObject)]
struct CreateQuestion {
    answer: String,
    question: String,
    quiz_template_id: String,
}

impl CreateQuestion {
    fn to_model(&self) -> Result<models::Question, oid::Error> {
        Ok(models::Question {
            answer: self.answer.clone(),
            id: ObjectId::new().to_string(),
            question: self.question.clone(),
            quiz_template_id: self.quiz_template_id.clone(),
        })
    }
}

#[derive(GraphQLInputObject)]
struct EditQuestion {
    answer: String,
    question: String,
    quiz_template_id: String,
}

impl EditQuestion {
    fn to_model(&self, id: &str) -> models::Question {
        models::Question {
            answer: self.answer.clone(),
            id: id.to_string(),
            question: self.question.clone(),
            quiz_template_id: self.quiz_template_id.clone(),
        }
    }
}

pub struct QuestionMutation;

#[graphql_object]
#[graphql(context = Context)]
impl QuestionMutation {
    async fn create(&self, context: &Context, input: CreateQuestion) -> FieldResult<Question> {
        let input_model = input.to_model()?;

        let model = context.questions.create(&input_model).await?;

        Ok(Question::from_model(&model))
    }

    async fn delete_by_id(&self, context: &Context, id: ID) -> FieldResult<bool> {
        let result = context.questions.delete_by_id(&id.to_string()).await?;

        Ok(result)
    }

    async fn edit(
        &self,
        context: &Context,
        id: ID,
        input: EditQuestion,
    ) -> FieldResult<bool> {
        let input_model = input.to_model(&id.to_string());

        let result = context.questions.update_by_id(&id.to_string(), &input_model).await?;
        
        Ok(result)
    }
}

pub struct QuestionQuery;

#[graphql_object]
#[graphql(context = Context)]
impl QuestionQuery {
    async fn all(&self, context: &Context) -> FieldResult<Vec<Question>> {
        let models = context.questions.read_all().await?;

        Ok(models.iter().map(Question::from_model).collect())
    }

    async fn by_id(&self, context: &Context, id: ID) -> FieldResult<Option<Question>> {
        let model = context.questions.read_by_id(&id.to_string()).await?;

        Ok(match model {
            Some(model) => Some(Question::from_model(&model)),
            None => None,
        })
    }
}

File: api/app/resolvers/question.py
from bson import ObjectId
from typing import TYPE_CHECKING, Annotated
from strawberry import ID, Private, field, input, lazy, type

from ..context import get_context
from ..models import QuestionModel, Questions

if TYPE_CHECKING:
    from .quiz_template import QuizTemplate


def get_questions() -> Questions:
    return get_context().questions


@type
class Question:
    answer: str
    id: ID
    question: str
    quiz_template_id: Private[ID]

    @staticmethod
    def from_model(model: QuestionModel) -> "Question":
        return Question(
            answer=model["answer"],
            id=ID(str(model["_id"])),
            question=model["question"],
            quiz_template_id=ID(str(model["quiz_template_id"])),
        )

    @field
    def quiz_template(self) -> Annotated["QuizTemplate", lazy(".quiz_template")]:
        from .quiz_template import QuizTemplate

        return QuizTemplate.from_model(
            get_context().quiz_templates.get_by_id(ObjectId(str(self.quiz_template_id)))
        )


@input
class CreateQuestion:
    answer: str
    question: str
    quiz_template_id: ID


@input
class EditQuestion(CreateQuestion):
    id: ID


@type
class QuestionMutation:
    @field
    def create(self, input: CreateQuestion) -> Question:
        return Question.from_model(
            get_questions().create(
                {
                    "answer": input.answer,
                    "question": input.question,
                    "quiz_template_id": ObjectId(str(input.quiz_template_id)),
                }
            )
        )

    @field
    def delete_by_id(self, id: ID) -> bool:
        return get_questions().delete_by_id(ObjectId(str(id)))

    @field
    def edit(self, input: EditQuestion) -> Question:
        return Question.from_model(
            get_questions().edit_by_id(
                ObjectId(str(input.id)),
                {
                    "answer": input.answer,
                    "question": input.question,
                    "quiz_template_id": ObjectId(str(input.quiz_template_id)),
                },
            )
        )


@type
class QuestionQuery:
    @field
    def all(self) -> list[Question]:
        return list(map(Question.from_model, get_questions().all()))

    @field
    def by_id(self, id: ID) -> Question:
        return Question.from_model(get_questions().get_by_id(ObjectId(str(id))))
